src/util/taproot.rs

```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+}
+
+impl TaprootSpendInfo {
+    /// Create a new [`TaprootSpendInfo`] from a list of script(with default script version) and
+    /// weights of satisfaction for that script. The weights represent the probability of
+    /// each branch being taken. If probabilities/weights for each condition are known,
+    /// constructing the tree as a Huffman tree is the optimal way to minimize average
+    /// case satisfaction cost. This function takes input a list of tuple(usize, &Script)
+    /// where usize represents the satisfaction weights of the branch.
+    /// For example, [(3, S1), (2, S2), (5, S3)] would construct a TapTree that has optimal
+    /// satisfaction weight when probability for S1 is 30%, S2 is 20% and S3 is 50%.
+    /// # Errors:
+    ///
+    /// - When the optimal huffman tree has a depth more than 128
+    /// - If the provided list of script weights is empty
+    pub fn new_huffman_tree<C, I>(
+        secp: &Secp256k1<C>,
+        internal_key: schnorr::PublicKey,
+        script_weights: I,
+    ) -> Result<Self, TaprootBuilderError>
+    where
+        I: IntoIterator<Item = (usize, Script)>,
+        C: secp256k1::Verification,
+    {
+        let mut node_weights = BinaryHeap::<(usize, NodeInfo)>::new();
+        for (p, leaf) in script_weights {
+            node_weights.push((p, NodeInfo::new_leaf_with_ver(leaf, LeafVersion::default())));
+        }
+        if node_weights.is_empty() {
+            return Err(TaprootBuilderError::IncompleteTree);
+        }
+        while node_weights.len() > 1 {
+            // Combine the last two elements and insert a new node
+            let (p1, s1) = node_weights.pop().expect("len must be atleast two");
+            let (p2, s2) = node_weights.pop().expect("len must be atleast two");
+            // Insert the sum of first two in the tree as a new node
+            node_weights.push((p1 + p2, NodeInfo::combine(s1, s2)?));
+        }
+        // Every iteration of the loop reduces the node_weights.len() by exactly 1
+        // Therefore, the loop will eventually terminate with exactly 1 element
+        assert!(node_weights.len() == 1);
+        let node = node_weights.pop().unwrap().1;
+        return Ok(Self::from_node_info(secp, internal_key, node));
+    }
+
+    /// Create a new key spend with internal key and proided merkle root.
+    /// Provide [`None`] for merkle_root if there is no script path.
+    /// # *Note*: As per BIP341
+    ///
+    /// When the merkle root is [`None`], the output key commits to an unspendable
+    /// script path instead of having no script path. This is achieved by computing
+    /// the output key point as Q = P + int(hashTapTweak(bytes(P)))G.
+    /// See also [`TaprootSpendInfo::tap_tweak`].
+    /// Refer to BIP 341 footnote (Why should the output key always have
+    /// a taproot commitment, even if there is no script path?) for more details
+    ///
+    pub fn new_key_spend<C: secp256k1::Verification>(
```

Comment:
Rust API recommends to name such functions `with_key_spend`

Reply:
I am still not super comfortable with `with_stuff` vs `new_stuff`. Plus, we already all APIs in rust-bitcoin are with `new_`, 

In this particular case, I think new_key_spend makes more sense vs `with_key_spend`. I think `with` makes more sense for simpler operations that directly operate on a field like `Vec::with_capacity()`. Maybe `with_internal_key`? I somewhat prefer `new_key_spend`, but I will go with what most reviewers/users want.

Reply:
`with_internal_key` seems more reasonable, but I will not ask to change just a function name since this is nit

Reply:
I don't have a strong opinion on this, but FWIW I think the correct thing is `with_` and I might consider fixing all the other instances of `new_` in a future mega-breaking API modernization PR.